simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java

```diff
@@ -448,25 +470,172 @@ public double getValue(Statistics statistics) {
     );
   }
 
+  private List<MetricFamilySamples> getPerQueryMetrics() {
+    List<MetricFamilySamples> metrics = new ArrayList<MetricFamilySamples>();
+
+    metrics.addAll(Arrays.asList(
+
+        createCounterForQuery("hibernate_per_query_cache_hit_total",
+                "Global number of cache hits for query (getCacheHitCount)",
+            new ValueProviderPerQuery() {
+              @Override
+              public double getValue(Statistics statistics, String query) {
+                return statistics.getQueryStatistics(query)
+                        .getCacheHitCount();
+              }
+            }
+        ),
+        createCounterForQuery("hibernate_per_query_cache_miss_total",
+                "Global number of cache misses for query (getCacheMissCount)",
+            new ValueProviderPerQuery() {
+              @Override
+              public double getValue(Statistics statistics, String query) {
+                return statistics.getQueryStatistics(query)
+                        .getCacheMissCount();
+              }
+            }
+        ),
+        createCounterForQuery("hibernate_per_query_cache_put_total",
+                "Global number of cache puts for query (getCachePutCount)",
+            new ValueProviderPerQuery() {
+              @Override
+              public double getValue(Statistics statistics, String query) {
+                return statistics.getQueryStatistics(query)
+                        .getCachePutCount();
+              }
+            }
+        ),
+        createCounterForQuery("hibernate_per_query_executions_total",
+                "Global number of executions for query (getExecutionCount)",
+            new ValueProviderPerQuery() {
+              @Override
+              public double getValue(Statistics statistics, String query) {
+                return statistics.getQueryStatistics(query)
+                        .getExecutionCount();
+              }
+            }
+        ),
+        createCounterForQuery("hibernate_per_query_execution_rows_total",
+                "Global number of rows for all executions of query (getExecutionRowCount)",
+            new ValueProviderPerQuery() {
+              @Override
+              public double getValue(Statistics statistics, String query) {
+                return statistics.getQueryStatistics(query)
+                        .getExecutionRowCount();
+              }
+            }
+        ),
+        createGaugeForQuery("hibernate_per_query_execution_min_time",
+                "Minimum execution time of query (getExecutionMinTime)",
```

Comment:
I would have liked counters better, too, but Hibernate's [QueryStatistics](https://docs.jboss.org/hibernate/core/3.3/api/org/hibernate/stat/QueryStatistics.html) only provides the current average, max and min values for execution times of single queries. 

Sadly, Hibernate provides no hook to count things...

However, for the use case of making especially long-running database queries transparent, the "max" metric is still helpful. We had the queries listed in a table in Grafana, sorted by their "max" value, which was quite enlightening :)

Reply:
In this situation we usually take only the average, and document what is actually is. How is this average implemented?

Reply:
The max, avg and min times are measured in milliseconds. Max and min are held as local variables over the lifetime of the application. The avg is calculated on access by dividing the accumulated time by the number of query executions (both also held as local variables).

[This](https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/main/java/org/hibernate/stat/internal/QueryStatisticsImpl.java) is the source.

Reply:
If you expose totalExecutionTime then you can calculate average on the Prometheus side, as with getExecutionCount it makes a Summary.

Reply:
I added the totalExecutionTime metric. I had to bump up the Hibernate dependency, though, because this metric has only been added with Hibernate 5 (which is three years old by now).